tests/test_views_admin.py

```diff
@@ -0,0 +1,242 @@
+
+    def test_cannot_delete_user_without_permission(self):
+        # Only the user with self.headers can access this URL.
+        self.app.delete(self.delete_user_url, headers=self.doomed_user, status=403)
+
+    def test_event_emitted_for_bucket_deletion(self):
+        delete_events = [
+            e for e in self.events if e.payload["action"] == events.ACTIONS.DELETE.value
+        ]
+        bucket_events = [e for e in delete_events if e.payload["resource_name"] == "bucket"]
+        # There should only be one such event, because all buckets
+        # share a parent
+        self.assertEqual(len(bucket_events), 1)
+        # There should only be one such bucket
+        self.assertEqual(len(bucket_events[0].impacted_objects), 1)
+        event = bucket_events[0].impacted_objects[0]
+        self.assertEqual(event["old"]["id"], "user_to_delete")
+
+    def test_event_not_emitted_for_coalesced_collection_deletion(self):
+        # Verify that collection deletion for
+        # /buckets/user_to_delete/collections/collection_to_delete
+        # was folded into bucket deletion for
+        # /buckets/user_to_delete
+        delete_events = [
+            e for e in self.events if e.payload["action"] == events.ACTIONS.DELETE.value
+        ]
+        collection_events = [
+            e for e in delete_events if e.payload["resource_name"] == "collection"
+        ]
+        # There should only be one such event, for /buckets/user_safe
+        self.assertEqual(len(collection_events), 1)
+        self.assertEqual(collection_events[0].payload["bucket_id"], "user_safe")
+        # Nothing for user_to_delete, so it must have been coalesced
+
+    def test_event_emitted_for_record_deletion(self):
+        delete_events = [
+            e for e in self.events if e.payload["action"] == events.ACTIONS.DELETE.value
+        ]
+        record_events = [e for e in delete_events if e.payload["resource_name"] == "record"]
+        # Just one, for safe_collection_doomed_record_url
+        self.assertEqual(len(record_events), 1)
+        event = record_events[0]
+        self.assertEqual(event.payload["bucket_id"], "user_safe")
+        self.assertEqual(event.payload["collection_id"], "collection_safe")
+        # Just the doomed record, not the safe one
+        self.assertEqual(len(event.impacted_objects), 1)
+        self.assertEqual(event.impacted_objects[0]["old"]["id"], "record_to_delete")
+
+
+class GetParentUriTest(unittest.TestCase):
+    def test_parent_uri_behaves_sensibly_for_unknown_resources(self):
+        unknown_url = "/cities/prague/monuments/castle/foods/mushrooms"
+        self.assertEqual(get_parent_uri(unknown_url), "/cities/prague/monuments/castle")
+
+    def test_parent_uri_accepts_pathological_urls(self):
+        # This shouldn't ever actually happen, and if it does happen
+        # it's probably a sign of a larger incompatibility of this
+        # method with the URL scheme
```

Comment:
If we try to support `/buckets`, then we should also try to support `/buckets/abcd/collections`. The method (as scoped right now) doesn't have to care too much about these kinds of URLs because it only considers actual object IDs.